internal/pkg/cacheutil/cacheserver.go

```diff
@@ -208,23 +203,16 @@ func (cs *CacheServer) Put(incomingData incoming.DataTypeInterface) {
 
 }
 
-func (cs CacheServer) close() {
-	quitCacheServerCh <- true
-	close(quitCacheServerCh)
-}
-
 func (cs CacheServer) loop() {
-LOOP:
+	debugc("Debug:CacheServer loop started")
 	for {
 		// Reuse buffer if there's room.
 		buffer := <-cs.ch
 		shard := cs.cache.GetShard(buffer.data.GetKey())
 		shard.SetData(buffer.data)
 		select {
 		case freeList <- buffer:
-		// Buffer on free list; nothing more to do.
-		case <-quitCacheServerCh:
-			break LOOP
+			// Buffer on free list; nothing more to do.
```

Comment:
This is a bit weird to me, will need to look closer and what this does. Basically what we have here is an attempt to read from buffer into freeList, and if either there is no message in buffer, or no space in freelist, we just carry on. Naively, this seems like an odd thing to do, but I haven't looked any deeper.

The mechanism for exiting this main cacheserver loop was never implmented (nothing ever calls close()) and I don't really see why you'd want this to shut down during operation.

Reply:
So I'm not that skilled in Go to give my hand to fire for the following senstence, but: CacheServer.loop is executed in separate goroutine and as such should be explicitly ended so we won't leave zombie threads after the smart gateway ends (either by accident or on purpose). So IMO the fact that the close method is not called is the actual bug :).

Reply:
That said we need to defer close instead.

Reply:
I'm not really worried about leaking threads, this loop runs for the lifetime of the process and when the parent ends, so do it's threads (our containers DO shut down), that said, I think you're totally right and a deferred close is proper. I was not attempting to alter or fix any functionality with this PR, just address the staticcheck errors, but this seems small enough that I could add it.

Reply:
Is this really a concern? NewCacheServer() is called just once from StartMetrics(), and it spawns the goroutine that runs this infinite loop. When StartMetrics() returns, main() will exit (https://github.com/redhat-service-assurance/smart-gateway/blob/master/cmd/main.go#L19). When the main goroutine exits, so do all other goroutines (https://golang.org/ref/spec#Program_execution) this happens abruptly without calling defers in the other goroutines. They do not leak.

Zombies processes are when a child exits and the parent does not wait on it to consume (or ignore) it's exit code; in terms of a pattern, that's actually the opposite of what's happening here! Though it's worth mentioning that there are no "parent" and "child" goroutines anyways, and goroutines are not processes (or threads), so they do not follow the unix process pattern of needing a wait() or equivalent SIGCHLD handling.

Functionally speaking, the smart gateway container exits when we tell it to, so I don't think there is a bug.

Our containers DO shut down when we tell them to.

Reply:
Goroutines runtimes are scheduled on OS threads IIUC, but you're right ... still it does not mean that it is not an antipattern :) [1][2], 

[1] https://dave.cheney.net/2016/12/22/never-start-a-goroutine-without-knowing-how-it-will-stop
[2] https://stackoverflow.com/questions/25518531/ok-to-exit-program-with-active-goroutine